addon/-record-data-private/system/model/internal-model.js

```diff
@@ -315,24 +318,35 @@ export default class InternalModel {
     this._doNotDestroy = false;
 
     if (this._record) {
-      Object.keys(this._relationshipPromisesCache).forEach(key => {
-        // TODO Igor cleanup the guard
-        // TODO there is probably relationship cleanup to do outside of the _record check
-        if (this._relationshipPromisesCache[key].destroy) {
-          this._relationshipPromisesCache[key].destroy();
-        }
-        delete this._relationshipPromisesCache[key];
-      });
-      Object.keys(this._manyArrayCache).forEach(key => {
-        this._retainedManyArrayCache[key] = this._manyArrayCache[key];
-        delete this._manyArrayCache[key];
-      });
       this._record.destroy();
     }
 
+    Object.keys(this._relationshipPromisesCache).forEach(key => {
+      // TODO Igor cleanup the guard
+      if (this._relationshipPromisesCache[key].destroy) {
+        this._relationshipPromisesCache[key].destroy();
+      }
+      delete this._relationshipPromisesCache[key];
+    });
+    // if the manyArray is for a sync relationship
+    //   we should clear it
+    Object.keys(this._manyArrayCache).forEach(key => {
+      let manyArray = (this._retainedManyArrayCache[key] = this._manyArrayCache[key]);
+      delete this._manyArrayCache[key];
+
+      /*
+          It is likely in this case instead of retaining we should destroy
+
+          - @runspired
+        */
+      if (manyArray && !manyArray._inverseIsAsync) {
+        manyArray.clear();
+      }
+    });
+
```

Comment:
Under what circumstances do we have entries in `_relationshipPromisesCache` or `_manyArrayCache` but no record?

Reply:
I suspect that should be an invalid state as of now but wouldn't necessarily be so in the future, moved back within the guard for `hasRecord`.